src/buildout/wheel/__init__.py

```diff
@@ -1,56 +1,119 @@
-import distlib.wheel
-import humpty
 import os
+import os.path
+import shutil
+import sys
 import pkg_resources
 import setuptools.package_index
 from six.moves.urllib import parse
 import wheel.install
+import wheel.decorator
 import pip.pep425tags
 import zc.buildout.easy_install
+import distutils.command.install
 
-orig_interpret_distro_name = setuptools.package_index.interpret_distro_name
+NAMESPACE_STUB_PATH = os.path.join(os.path.dirname(__file__), 'namespace_stub.py')
+assert os.path.isfile(NAMESPACE_STUB_PATH)
+
+orig_distros_for_location = setuptools.package_index.distros_for_location
 original_wheel_to_egg = zc.buildout.easy_install.wheel_to_egg
 
-wheel_supported = pip.pep425tags.get_supported
+class SelfDestructingEgg(pkg_resources.Distribution):
+    """Since the location for creating the egg is inside the download directory
+    We must self destruct to avoid having both the wheel and the generated egg
+    there at the same time."""
+    def __del__(self):
+        shutil.rmtree(self.location)
+
+class WheelFile(wheel.install.WheelFile):
+
+    def __init__(self, *args, **kwargs):
+        # use pip's notion of supported tags, not wheel's:
+        kwargs.setdefault('context', pip.pep425tags.get_supported)
+        super(WheelFile, self).__init__(*args, **kwargs)
+
+    def lay_egg_into(self, target):
+        join = os.path.join
+        # drop everything inside the metadata directory, keyed by section name,
+        # except the code itself
+        location = join(target, self.egg_name)
+        overrides = {
+            key: (location if key in ('platlib', 'purelib')
+                    else join(location, self.distinfo_name, key))
+            for key in distutils.command.install.SCHEME_KEYS
+        }
+        self.install(overrides=overrides)
+        # pkg_resources special-cases directories in sys.path with `.egg`
+        # extension and expect their metadata directory to be called EGG-INFO
+        # instead of the `[name]-[version].dist-info` directory of wheels.
+        egg_info_location = join(location, 'EGG-INFO')
+        os.rename(join(location, self.distinfo_name), egg_info_location)
+        # See docstring of SelfDestructingEgg above for why it's needed:
+        metadata = pkg_resources.PathMetadata(location, egg_info_location)
+        egg = self.distribution(location, metadata=metadata,
+                                class_=SelfDestructingEgg)
+        # Fix namespaces with missing __init__
+        if (sys.version_info < (3, 3) and
```

Comment:
I was about to comment that "nobody supports Python 3.2 any more, why do we care", and then I realized this condition also matches Python 2.7.